src/examples/libpmem2/pmemset/pmemset.h

```diff
@@ -0,0 +1,357 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/* Copyright 2020, Intel Corporation */
+
+#ifndef LIBPMEMSET_H
+#define LIBPMEMSET_H 1
+
+#include <stddef.h>
+#include <stdint.h>
+#include <sys/types.h>
+#include <libpmem2.h>
+
+struct pmemset;
+struct pmemset_config;
+struct pmemset_source;
+struct pmemset_part;
+struct pmemset_part_map;
+
+int pmemset_config_new(struct pmemset_config **config);
+
+/*
+ * By default, the library only attempts to open the parts. These two functions
+ * modify this behavior.
+ * Create if none behaves like O_CREAT, but also initializes part metadata.
+ * Create if invalid checks if the existing file contains a valid pool,
+ * and if not, creates a new one (this is useful for close-to-open).
+ */
+void pmemset_config_set_create_if_none(struct pmemset_config *config,
+	int value);
+void pmemset_config_set_create_if_invalid(struct pmemset_config *config,
+	int value);
+
+/*
+ * The synchronous event stream. These events are generated inside of the
+ * flush/memcpy functions on the set.
+ */
+enum pmemset_event {
+	PMEMSET_EVENT_COPY,
+	PMEMSET_EVENT_FLUSH,
+	PMEMSET_EVENT_DRAIN,
+	PMEMSET_EVENT_PERSIST,
+	PMEMSET_EVENT_BAD_BLOCK,
+	PMEMSET_EVENT_PART_ADD,
+	PMEMSET_EVENT_PART_REMOVE,
+};
+
+struct pmemset_event_copy {
+	void *addr;
+	size_t len;
+};
+
+struct pmemset_event_flush {
+	void *addr;
+	size_t len;
+};
+
+struct pmemset_event_drain {
+};
+
+struct pmemset_event_persist {
+	void *addr;
+	size_t len;
+};
+
+struct pmemset_event_bad_block {
+	void *addr;
+	size_t len;
+};
+
+struct pmemset_event_part_remove {
+	void *addr;
+	size_t len;
+	int fd;
+};
+
+struct pmemset_event_part_add {
+	void *addr;
+	size_t len;
+	int fd;
+};
+
+struct pmemset_event_context {
+	enum pmemset_event type;
+	union {
+		char _data[64];
+		struct pmemset_event_copy copy;
+		struct pmemset_event_flush flush;
+		struct pmemset_event_drain drain;
+		struct pmemset_event_persist persist;
+		struct pmemset_event_bad_block bad_block;
+		struct pmemset_event_part_remove part_remove;
+		struct pmemset_event_part_add part_add;
+	} data;
+};
+
+/*
+ * This callback can be used to create a copy of the data or directly
+ * replicate it somewhere. This is *not* an append-only log, nor is the
+ * data versioned in any way. Once the function exits, the memory range
+ * can no longer be accessed.
+ * There's no guarantee that accessing the data inside of the callback
+ * is thread-safe. The library user must guarantee this by not
+ * having multiple threads mutating the same region on the set.
+ */
+typedef int pmemset_event_callback(struct pmemset *set,
+	struct pmemset_event_context *context, void *arg);
+
+void pmemset_config_set_event_callback(struct pmemset_config *config,
+	pmemset_event_callback *callback, void *arg);
+
+/*
+ * Sets the address space reservation that will be used by the pmemset.
+ */
+void pmemset_config_set_reservation(struct pmemset_config *config,
```

Comment:
this needs a discussion of how it will be used. (especially for parts not placed continuously )